feature/CRUD_operations_for_courses#315

adeel-ahmad-awan opened the pull request:
### **Summary of Changes**

**Ticket:** [TESS-315](https://app.zenhub.com/workspaces/th-developmentdesigning-board-6565ea6f792dae0625b5c739/issues/gh/scilifelab-traininghub-platform/tess/315)

#### **Implemented Features**

* Added full **CRUD functionality** for *Courses*, including:
  * Complete model with necessary **associations** and **validations**
  * Controller with all standard **actions** and **authorization**
  * Views for creating, editing, viewing, and listing courses
    * Views include **breadcrumbs** for improved navigation
* Added **test cases** for both the model and controller

#### **Next Steps / To-Do**
* Add **JSON API** for the course catalogue
* Add **Report** section on the course *show* page
* Add **Activity Log** on the course *show* page
* Implement **permission checks** based on `TeSS::Config.feature['courses']`
* Create new task as per feedback, eg orcid automation
* Implement course email automation (e.g., determine when automatic emails should be triggered in the course flow)
* Add integration tests once the full feature, including the email automation, is deployed and stable.
* (Unrelated to the above) Add integration tests for the event workflow, covering event creation, event visibility, and email notifications within the event workflow.


---

## Course Feature – Test Plan

### 1. Controller Tests (`test/controllers/courses_controller_test.rb`)

#### CRUD Actions
##### Create
- Verify that a new course can be created successfully.
- Ensure all mandatory fields are validated (e.g., title, description, etc.).
- Test creation with missing or invalid fields to confirm:
  - Appropriate validation errors.
  - Correct UI feedback and layout.

##### Edit / Update
- Verify that an existing course can be edited and updated.
- Confirm validation works for both valid and invalid updates.
- Check the UI layout for:
  - Very short input values.
  - Very long input values (ensure layout remains intact).

##### Show
- Confirm the course details page loads successfully.
- Verify all associated data (e.g., events, nodes, contributors) displays correctly.

##### Index
- Ensure newly created courses appear in the course list.
- Test that:
  - All filters work correctly (e.g., by node, content provider, user).
  - Search functionality returns relevant results.

##### Destroy
- Verify that a course can be deleted successfully.
- Confirm dependent associations (like events) are handled as expected (`dependent: :nullify`).

### 2. Model Tests (`test/models/course_test.rb`)

#### Associations
Ensure all associations function correctly:
- `Course` → `has_many :events`
- `Course` → `has_and_belongs_to_many :nodes`
- `Course` → `has_and_belongs_to_many :content_providers`
- `Course` → `belongs_to :user`

Test that:
- Associations persist correctly in the database.
- Deleting or updating a course behaves correctly (e.g., nullifies or preserves relationships).
- Associated data displays properly in the UI.

### 3. Validation Tests
- Confirm all field-level validations:
  - Required fields are enforced.
  - Length and format constraints are respected.
- Test boundary inputs:
  - Very short values.
  - Very long values.
- Ensure the UI remains consistent and visually stable in all cases.

### 4. Authors & Contributors
- Verify that authors and contributors are saved in the correct format.
- When data is missing or invalid:
  - Correct validation errors are displayed.
  - Error messages are clear and helpful.

### 5. Help Texts & Form Labels
- Check that all help texts and tooltips in forms and pages are:
  - Present.
  - Accurate.
  - Helpful to the user.

### 6. Field Completeness
- Verify that all fields discussed for the Course model and views are:
  - Present.
  - Properly displayed.
  - Editable where applicable.

### 7. JSON API Tests
Test both **index** and **show** endpoints:
- Confirm that courses are available in JSON format.
- Ensure all relevant fields are included in the response.
- Validate that JSON structure and keys match the expected schema.

### 8. Test Coverage Review
- Run all tests:
  - `test/controllers/courses_controller_test.rb`
  - `test/models/course_test.rb`
- Confirm coverage includes:
  - CRUD operations
  - Associations
  - Validations
  - JSON responses
  - UI behavior and feedback
- Ensure all critical scenarios for the **first release** are fully tested and passing.

Commit:
feature/CRUD_operations_for_courses#315

harshita3199 committed:
add architecture diagram (#340)

Signed-off-by: Harshita Gupta <harshita.gupta@scilifelab.uu.se>
Co-authored-by: GeneralDido <dimpanouris@gmail.com>

harshita3199 committed:
update readme with architecture overview (#343)

Signed-off-by: Harshita Gupta <harshita.gupta@scilifelab.uu.se>

Commit:
feature/working CRUD for courses#315

kristen-schroeder commented:
@adeel-ahmad-awan  
feedback:
- there should be some way of indicating contact info for the course.
ie - contact info for the nbis courses is intro-to-ml@nbis.se, separated from individual author info

- events should not be mandatory. some courses run every other year.

- is there some link to the username of the author when the contact details are input? can I select ie, Nina if I'm adding her? Zenodo and Figshare have searchable fields where you can input orcid and it pulls the deets. the hiC course had 21 authors and 3 contributors, please help admin life be easy :D

- suggestion for mandatory fields: title, url, language, description, target audience, prereqs, structure & duration, LOs, content providers, scilifelab affiliation 
these being mandatory makes it consistent with courseInstance.

looks good so far!

Commit:
feature/ Working CRUD operations for courses #315

Commit:
added form validation for the course create and edit#315

Commit:
added delete dependency to course for the course delete#315

adeel-ahmad-awan commented:
[Issues - Detail.pdf](https://github.com/user-attachments/files/22927003/Issues.-.Detail.pdf)
review/remarks from @GeneralDido

Commit:
added update logic for authors and contributors #315

GeneralDido committed:
Fix:Update assets link (#344)

harshita3199 committed:
integrate matomo analytics  (#338)

Signed-off-by: Harshita Gupta <harshita.gupta@scilifelab.uu.se>
Co-authored-by: GeneralDido <dimpanouris@gmail.com>

GeneralDido committed:
Feature: Course Instance Redesign (#336)

Co-authored-by: kristen-schroeder <kristen.schroeder@scilifelab.se>

Commit:
refactor and fixed code review #315

kristen-schroeder committed:
kristen/updating hint fields for course CRUD (#351)

Commit:
#315 | Added test cases for model and controller in-progress

Commit:
#315 | Added test cases for model and controller

Commit:
added course in event form and remove useless code

GeneralDido committed:
Feature: Redesign course catalogue (#354)

Signed-off-by: Harshita Gupta <harshita.gupta@scilifelab.uu.se>
Co-authored-by: Harshita Gupta <39816522+harshita3199@users.noreply.github.com>
Co-authored-by: kristen-schroeder <kristen.schroeder@scilifelab.se>

GeneralDido committed:
Merge branch 'rc-1.5.0' into feature/CRUD_operations_for_courses#315

Commit:
api bug removed

GeneralDido committed:
Feature/Connect Course and Course Instance (#373)

GeneralDido committed:
Merge branch 'rc-1.5.0' into feature/CRUD_operations_for_courses#315

GeneralDido committed:
fix: add preload_index_associations under CoursesController

Commit:
Add integration tests for events #315

GeneralDido committed:
Harden check_exists and stabilize course/event forms

Hardens check_exists across resources (minimal JSON + non-disclosure + duplicate-safe selection), restores robust course prefill with inline errors, fixes Turbolinks double-
    binding for course authors/contributors, improves cookie consent compatibility + test stability, and includes small tracking/perf cleanups

GeneralDido committed:
Use course-style cards on events index

Switches the /events List tab to render events using a new course-card-style partial, matching /courses

Commit:
refactored logic for sending emails and slack notification for event workflow #315

Commit:
added email workflow for course, refactor email workflow for event and course

Commit:
added integration test for the course module workflow and resolve minor bugs#315

GeneralDido committed:
Add Courses dropdown nav and update headings

Replace separate Courses/Catalogue tabs with a single Courses dropdown linking to /courses (“Course Catalogue”) and /events (“Upcoming Courses”); update /courses
    heading to use i18n; set features.events.long to “Upcoming Courses”; remove visited/active-looking styling in the dropdown.

GeneralDido committed:
Show review-status icons on course/event cards

Add shared review_status_icon(record) helper (handles event_status/course_status, including approved) and render it on both new index card partials; add matching
    CSS colors.

GeneralDido committed:
Clean up Slack notifier and fix tests

Make Slack published notifier consistently use record; remove debug pp from course_filter; adjust controller tests to set created courses to approved so they
    hit authorization logic (not 404); fix the event Slack enqueue test setup.

Commit:
refactored the logic for acces of ownershor for course and event show update and edit page

Commit:
resolved small bug

harshita3199 commented on app/controllers/content_providers_controller.rb:
```diff
@@ -36,16 +36,20 @@ def edit
   # POST /events/check_exists.json
   def check_exists
-    @content_provider = ContentProvider.check_exists(content_provider_params)
+    @content_provider = ContentProvider.check_exists_candidates(content_provider_params)
```
Is this change relevant to this ticket?? I am not able to understand. @GeneralDido

GeneralDido replied:
We should keep, it's a good optimization regarding the URL checker:
- It returns the minimal payload instead of the whole object. 
- It also prevents possible leaks, for example suppose you have the same URL in another unverified course, previously it would do `ContentProvider.check_exists(...)`, which would return the first object. Now it limit orders them, and skips the ones from unapproved.

harshita3199 replied:
then we keep it!

GeneralDido committed:
prevent course instances for unapproved courses

Adds server-side guards to stop linking/approving Events against unapproved Courses; updates course/event dropdown logic, blocks “Create Course Instance” on unapproved Course pages, and shows status icons on instance tiles (plus tests).

GeneralDido committed:
block unapproving courses with approved instances and update locals

harshita3199 commented on app/controllers/courses_controller.rb:
```diff
@@ -0,0 +1,214 @@
+  def check_exists
+    @course = Course.check_exists_candidates(course_check_exists_params)
+                    .limit(50)
+                    .find { |course| course_disclosable_for_check_exists?(course) }
```
Similar to previous comment??

GeneralDido replied:
Yep, same, we limit, then get the first correct one, it also handles duplicates (so if you have a duplicate URL it's gonna show up)

harshita3199 commented on app/controllers/courses_controller.rb:
```diff
@@ -0,0 +1,214 @@
+      { keywords: [] },
+      { target_audience: [] },
+      { authors: [:name, :affiliation, :orcid, :email] },
+      { contributors: [:name, :affiliation, :orcid, :email] },
```
Email needs to be removed here in contributor

harshita3199 commented on app/controllers/courses_controller.rb:
```diff
@@ -0,0 +1,219 @@
+    end
+  end
+
+  def preload_index_associations
```
What does this function do?

GeneralDido replied:
Optimization, it preloads the courses + data we need, so in the courses index page we don't have to do multiple queries (to load the course and per course to load nodes, providers, events etc).

harshita3199 commented on app/controllers/events_controller.rb:
```diff
@@ -285,10 +299,24 @@ def set_event
   end
 
+  def event_disclosable_for_check_exists?(event)
+    return false unless policy(event).show?
```
policy??

GeneralDido replied:
check in `app/policies/` , in this case if the event should not be shown, for example if it's unverified, don't show it in the URL checker. 

Else, if it's allowed/verified, etc. you can see something like this:
```
Notice: This URL is already registered as: <Event URL>
```

harshita3199 commented on app/controllers/materials_controller.rb:
```diff
@@ -72,17 +72,21 @@ def edit
   # POST /materials/check_title.json
   def check_exists
-    @material = Material.check_exists(material_params)
+    @material = Material.check_exists_candidates(material_params)
```
not needed here for this PR

GeneralDido replied:
Keep, it's the same optimisation as above, I applied it for all (changed the basic one that was before). 
In any case we're going to refactor/change code when we get to materials anyway.

harshita3199 commented on app/controllers/sources_controller.rb:
```diff
@@ -55,17 +55,21 @@ def create
   # POST /sources/check_exists.json
   def check_exists
-    @source = Source.check_exists(source_params)
+    @source = Source.check_exists_candidates(source_params)
```
not needed for this PR

GeneralDido replied:
same as above

harshita3199 commented on app/models/event.rb:
```diff
@@ -566,8 +576,41 @@ def content_provider_ids=(ids)
     self.content_providers = ContentProvider.where(id: ids.reject(&:blank?))
   end
 
+  def event_status_just_approved?
```
duplicating with courses, could have one method and use it here as well

GeneralDido replied:
Correct, although if I remember correctly we had discussed this with Adeel and we came up with the conclusion to keep it separate because the two objects (course / course instance) are not necessary the same, and theoretically at one point we might need to change one of them, so it doesn't hurt in this case to have the duplicate. 

Although, tbf even in that case you can theoretically optimize with one function and change what you return for each case, but too much of a hustle right now imho.

harshita3199 commented on app/controllers/content_providers_controller.rb:
```diff
@@ -94,6 +98,16 @@ def destroy
 
   private
 
+  def content_provider_disclosable_for_check_exists?(content_provider)
```
Further discussion on how this works will be good to have

GeneralDido replied:
It's used in the URL checker in the Content Provider form: when you type a URL, the app tries to detect “a provider with this URL already exists” and (if allowed) returns {id,title} + a link.

- policy: uses `ContentProviderPolicy` , which inherits from `ScrapedResourcePolicy` (Adeel also wrote code there). The check basically states: if a content provider belongs to an unverified/rejected user, normal users should not be able to see it.
- shadowban: if we didn’t add this guard, a normal user could theoretically bypass “shadowban hides from search” by simply probing /check_exists with URLs. Very rare, but doesn't hurt to have it there.

harshita3199 commented on app/controllers/courses_controller.rb:
```diff
@@ -0,0 +1,219 @@
+
+  private
+
+  def course_disclosable_for_check_exists?(course)
```
similar to the previous comment for content_provider

GeneralDido replied:
same

harshita3199 commented on app/controllers/courses_controller.rb:
```diff
@@ -0,0 +1,219 @@
+    return false unless policy(course).show?
+
+    if course.respond_to?(:from_shadowbanned?) && course.from_shadowbanned?
+      return current_user&.shadowbanned? || current_user&.is_admin?
```
test shadowbanned feature in the portal

harshita3199 replied:
not tested this yet because course catalogue needs priority and as per Dimitris comment shadowbanned is good to have.

GeneralDido committed:
if the upcoming course instance is approved, do not show the icon

harshita3199 commented:
i have deployed the latest changes to pre-prod, tested them they are working as expected. Also, ran the integration tests which are working fine and unit tests are also passing so we are good to merge this PR.

Unit test output below:
`
1246 tests, 5241 assertions, 0 failures, 0 errors, 102 skips
Coverage report generated for Minitest to /code/coverage. 16592 / 20311 LOC (81.69%) covered.
Lcov style coverage report generated for Minitest to /code/coverage/lcov/code.lcov`

harshita3199 reviewed (approved):
See the last comment in the PR. https://github.com/SciLifeLab-TrainingHub-Platform/TeSS/pull/342#issuecomment-3834269718
